### ISSUE TITLE
Asyncrhonous code in the message processor operators.

### ISSUE THREAD

**Issue by RogerSep** [LGouellec/kafka-streams-dotnet#145]
Description
I'm not sure if I'm not reading correctly the API, but there are no variations for async code in the creation of the streams. I don't find examples or issues with async code.

How to reproduce
================
```
// AsyncStream
builder.Stream<string, string>("topic")
    .Foreach(async (k, v) =>
    {
      await Task.Delay(500); // Imagine a database call or something.
    });

// SyncStream
var syncStream = builder.Stream<string, string>("topic")
    .Foreach((k, v) =>
    {
      (Task.Delay(500)).ConfigureAwait(false).GetAwaiter().GetResult();
    });

// FilterStream
builder.Stream<string, string>("topic")
    .Filter(async (k, v) =>
    {
      await Task.Delay(500);
      return true;
    });
```
With these three streams samples, my observations are:
- `FilterStream` is not possible unless the operation is blocking.
- `AsyncStream` kind of works, but, it will process every single Message without waiting for the operations to be completed, essentially processing the messages in parallel in a fire and forget fashion.
- `SyncStream` works better in the sense that it will process messages in order, but I feel like this is not the solution.

I believe this is desired? as operations need to be synchronous. I understand this can slow the throughput of an application. Any thoughts? is there a recommended pattern to follow in these kind of situations?

Thanks!

Version of streamiz. 1.1.5

**Comment by LGouellec** [comment]
Hi @RogerSep,

Thanks for your issue.

Kafka streams process synchronously message by message before flushing state store and committing source offset. 

The loop processing model is :
- Consume messages from source topics
- Process messages
- Flush producer and state store
- Commit source offets 

So the processing phase have to execute synchronously. 

By the way, calling external service inside your topology is not a good practice. Mostly with JAVA implementation but also with .Net one.
FYI : https://dev.to/bleporini/why-you-should-not-query-a-database-in-your-stream-processors-4829. 

Best regards,

**Comment by LGouellec** [comment]
But maybe, we add async syntax with this workflow : 

- Consume messages from source topics
- Task.FromResult(Process messages async);
- Flush producer and state store
- Commit source offets

Because we have to wait end of processing before committing, else you could commit message before the end of processing, and in case of error loose messages (in processing term).

Kr,

**Comment by LGouellec** [comment]
Hi @RogerSep,

I created a new issue about External async call inside your topology. 

See #147

Kr,
